// tern-webclipse/code_folding.js
var walk = require("acorn/dist/walk");
var acorn = require("acorn/dist/acorn");

module.exports.computeCodeFolding = function(ast) {
	var elements = [];
	walk.simple(ast, {
		VariableDeclaration: function (node) {
			elements.push({type: "field", start: node.start, end: node.end});
		},
		MethodDefinition: function (node) {
			elements.push({type: "method", start: node.start, end: node.end});
		},
		Function: function (node) {
			elements.push({type: "function", start: node.start, end: node.end});
		},
		ClassDeclaration: function (node) {
			elements.push({type: "class", start: node.start, end: node.end});
		},
		ImportDeclaration: function (node) {
			elements.push({type: "import", start: node.start, end: node.end});
		},
		ExportNamedDeclaration: function (node) {
			elements.push({type: "export", start: node.start, end: node.end});
		},
		ExpressionStatement: function (node) {
			elements.push({type: "expression_statement", start: node.start, end: node.end});
		}
	})
	return {
		"elements": elements
	}	
}
